services/transactionService.ts

```diff
@@ -0,0 +1,554 @@
+  }
+  console.log('checkIfTransactionHasBeenSpeedup', {
+    recentDonationsLength: userRecentTransactions.length,
+    lastDonation: userRecentTransactions[userRecentTransactions.length - 1],
+    nonce,
+    page
+  })
+  const smallestNonce: number = userRecentTransactions[
+    userRecentTransactions.length - 1
+  ].nonce as number
+
+  if (smallestNonce < nonce) {
+    // because the list is descending if the nonce is greater than our desired nonce,
+    // the other transactions nonce will not match out transaction so we throw exception
+    throw new Error(errorMessages.TRANSACTION_NOT_FOUNT_IN_USER_HISTORY)
+  }
+
+  return checkIfTransactionHasBeenSpeedup({
+    input,
+    page: page + 1,
+    contractAddress
+  })
+}
+
+async function getTransactionDetailForNormalTransfer(
+  input: TransactionDetailInput
+): Promise<NetworkTransactionInfo> {
+  const { txHash, symbol, networkId, fromAddress, toAddress, nonce } = input
+  const transaction = await getNetworkWeb3(networkId).eth.getTransaction(txHash)
+  if (
+    transaction &&
+    transaction.from.toLowerCase() !== fromAddress.toLowerCase()
+  ) {
+    throw new Error(
+      errorMessages.TRANSACTION_FROM_ADDRESS_IS_DIFFERENT_FROM_SENT_FROM_ADDRESS
+    )
+  }
+  if (
+    transaction &&
+    transaction.to &&
+    transaction.to.toLowerCase() !== toAddress.toLowerCase()
+  ) {
+    throw new Error(
+      errorMessages.TRANSACTION_TO_ADDRESS_IS_DIFFERENT_FROM_SENT_TO_ADDRESS
+    )
+  }
+  // ADD checking for toAddress and amount
+
+  if (transaction) {
+    return {
+      from: transaction.from,
+      to: transaction.to as string,
+      hash: txHash,
+      amount: normalizeAmount(transaction.value, 18),
+      currency: symbol
+    }
+  }
+
+  if (!nonce) {
+    // if nonce didn't pass, we can not understand whether is speedup or not
+    throw new Error(errorMessages.TRANSACTION_NOT_FOUND)
+  }
+  return checkIfTransactionHasBeenSpeedup({ input })
+}
+
+async function getTransactionDetailForTokenTransfer(
+  input: TransactionDetailInput
+): Promise<NetworkTransactionInfo> {
+  const { txHash, symbol, networkId, fromAddress, nonce, toAddress } = input
+  const token = findTokenByNetworkAndSymbol(networkId, symbol)
+  const web3 = getNetworkWeb3(networkId)
+  const transaction = await web3.eth.getTransaction(txHash)
+  console.log('getTransactionDetailForTokenTransfer', { transaction, input })
+  if (
+    transaction &&
+    transaction.to?.toLowerCase() !== token.address.toLowerCase()
+  ) {
+    throw new Error(
+      errorMessages.TRANSACTION_SMART_CONTRACT_CONFLICTS_WITH_CURRENCY
+    )
+  }
+  if (
+    transaction &&
+    transaction.from.toLowerCase() !== fromAddress.toLowerCase()
+  ) {
+    console.log('fromAddress of donation is different', {
+      transaction,
+      input
+    })
+    throw new Error(
+      errorMessages.TRANSACTION_FROM_ADDRESS_IS_DIFFERENT_FROM_SENT_FROM_ADDRESS
+    )
+  }
+  if (transaction) {
```

Comment:
I suggest finding the transaction first (hash or nonce matches) and then doing the verification